temporal-sdk/src/main/java/io/temporal/worker/tuning/ResourceBasedController.java

```diff
@@ -0,0 +1,139 @@
+  private Instant lastPidRefresh = Instant.now();
+
+  private Scope metricsScope = new NoopScope();
+  private boolean metricsScopeSet = false;
+  private volatile double last_mem_usage = 0;
+  private volatile double last_cpu_usage = 0;
+  private volatile double last_mem_pid_out = 0;
+  private volatile double last_cpu_pid_out = 0;
+
+  /**
+   * Construct a controller with the given options. If you want to use resource-based tuning for all
+   * slot suppliers, prefer {@link ResourceBasedTuner}.
+   */
+  public static ResourceBasedController newSystemInfoController(
+      ResourceBasedControllerOptions options) {
+    return new ResourceBasedController(options, new JVMSystemResourceInfo());
+  }
+
+  /**
+   * Construct a controller with the given options and system info supplier. Users should prefer
+   * {@link #newSystemInfoController(ResourceBasedControllerOptions)}.
+   */
+  public ResourceBasedController(
+      ResourceBasedControllerOptions options, SystemResourceInfo systemInfoSupplier) {
+    this.options = options;
+    this.systemInfoSupplier = systemInfoSupplier;
+    this.memoryController =
+        new PIDController(
+            options.getTargetCPUUsage(),
+            options.getMemoryPGain(),
+            options.getMemoryIGain(),
+            options.getMemoryDGain());
+    this.cpuController =
+        new PIDController(
+            options.getTargetCPUUsage(),
+            options.getCpuPGain(),
+            options.getCpuIGain(),
+            options.getCpuDGain());
+  }
+
+  /**
+   * @return True if the PID controllers & and other constraints would allow another slot
+   */
+  boolean pidDecision() {
+    decisionLock.lock();
+    try {
+      double memoryUsage = systemInfoSupplier.getMemoryUsagePercent();
+      double cpuUsage = systemInfoSupplier.getCPUUsagePercent();
+      double memoryOutput =
+          memoryController.getOutput(lastPidRefresh.getEpochSecond(), memoryUsage);
+      double cpuOutput = cpuController.getOutput(lastPidRefresh.getEpochSecond(), cpuUsage);
+      lastPidRefresh = Instant.now();
+      last_mem_usage = memoryUsage;
+      last_cpu_usage = cpuUsage;
+      last_mem_pid_out = memoryOutput;
+      last_cpu_pid_out = cpuOutput;
+
+      return memoryOutput > options.getMemoryOutputThreshold()
+          && cpuOutput > options.getCpuOutputThreshold()
+          && canReserve();
+    } finally {
+      decisionLock.unlock();
+    }
+  }
+
+  private boolean canReserve() {
+    return systemInfoSupplier.getMemoryUsagePercent() < options.getTargetMemoryUsage();
+  }
+
+  /** Visible for internal usage. Can only be set once. */
+  public void setMetricsScope(Scope metricsScope) {
+    synchronized (this) {
+      if (metricsScopeSet) {
+        return;
+      }
+      this.metricsScope = metricsScope;
+      metricsScopeSet = true;
+      // Launch a task to periodically emit the metrics
+      ForkJoinPool.commonPool().execute(this::emitMetrics);
```

Comment:
Would rather not eat a thread here. Since Tally doesn't expose a pull model and I'm guessing you are unable to set the metrics scope at construction for whatever reason, I would recommend once this setter is called, you an `AtomicReference<Gauges>` where `Gauges` is a static inner class of the 4 gauges, and then `.get()` and if not null set them in the `pidDecision()` call when you get them. This should be cheap and doesn't eat a thread;.

Reply:
Sounds good. I had some concern about calling them overly frequently, but I suppose pid decision is already appropriately throttled anyway.